util/indexer/src/service.rs

```diff
@@ -0,0 +1,1641 @@
+                                        indexer.rollback().expect("rollback block should be OK");
+                                    }
+                                }
+                                None => {
+                                    error!("long fork detected, ckb-indexer stored block {}, ckb node returns none, please check if ckb-indexer is connected to the same network ckb node.", longest_fork_number);
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                    None => {
+                        break;
+                    }
+                }
+            } else {
+                match self.get_block_by_number(0) {
+                    Some(block) => indexer.append(&block).expect("append block should be OK"),
+                    None => {
+                        error!("ckb node returns an empty genesis block");
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    /// Processes that handle block cell and expect to be spawned to run in tokio runtime
+    pub fn spawn_poll(&self, notify_controller: NotifyController) {
+        let initial_service = self.clone();
+        let initial_syncing = self
+            .async_handle
+            .spawn_blocking(move || initial_service.try_loop_sync());
+        let mut stop = self.stop.clone();
+        let async_handle = self.async_handle.clone();
+        let poll_service = self.clone();
+        self.async_handle.spawn(async move {
+            let _initial_finished = initial_syncing.await;
+            let mut new_block_receiver = notify_controller
+                .subscribe_new_block(SUBSCRIBER_NAME.to_string())
+                .await;
+            let sleep = time::sleep(poll_service.poll_interval);
+            tokio::pin!(sleep);
+            loop {
+                tokio::select! {
+                    Some(_) = new_block_receiver.recv() => {
+                        let service = poll_service.clone();
+                        if let Err(e) = async_handle.spawn_blocking(move || {
+                            service.try_loop_sync()
+                        }).await {
+                            error!("ckb indexer syncing join error {:?}", e);
+                        }
+                    },
+                    _ = &mut sleep => {
+                        let service = poll_service.clone();
+                        if let Err(e) = async_handle.spawn_blocking(move || {
+                            service.try_loop_sync()
+                        }).await {
+                            error!("ckb indexer syncing join error {:?}", e);
+                        }
+                    }
+                    _ = stop.changed() => break,
+                }
+            }
+        });
+    }
+
+    fn get_block_by_number(&self, block_number: u64) -> Option<core::BlockView> {
```

Comment:
Since the entire block data was only available via rpc `get_block_by_number` in the standalone service, but now that it's merged together, we can optimize this fn to load block header and block body only since the `indexer#append` fn only requires the header and transactions. This helps reduce the number of opening secondary db column family and reduce the unnecessary loading of uncle/proposal id / etc.

Reply:
Sounds good  👍  https://github.com/nervosnetwork/ckb/pull/3609/commits/89da2b070428da4d47333c8d2f8247ffc44579c2